test/e2e/vmagent_test.go

```diff
@@ -0,0 +1,295 @@
+	skipIfNoVMs(t)
+	data, err := setupTest(t)
+	if err != nil {
+		t.Fatalf("Error when setting up test: %v", err)
+	}
+	defer teardownTest(t, data)
+	vmList, err := setupVMAgentTest(t, data)
+	if err != nil {
+		t.Fatalf("Error when setting up VMAgent test: %v", err)
+	}
+	defer teardownVMAgentTest(t, data, vmList)
+	t.Run("testExternalNode", func(t *testing.T) { testExternalNode(t, data, vmList) })
+}
+
+// setupVMAgentTest creates ExternalNode, starts antrea-agent
+// and returns a list of VMs upon success
+func setupVMAgentTest(t *testing.T, data *TestData) ([]vmInfo, error) {
+	t.Logf("List of Windows VMs: '%s', Linux VMs: '%s'", testOptions.windowsVMs, testOptions.linuxVMs)
+	t.Logf("Using ServiceAccount %s, NameSpace %s", serviceAccount, namespace)
+	var vmList []vmInfo
+	if testOptions.linuxVMs != "" {
+		vms := strings.Split(testOptions.linuxVMs, ",")
+		for _, vm := range vms {
+			t.Logf("Get info for Linux VM: %s", vm)
+			tempVM := getVMInfo(t, data, vm)
+			vmList = append(vmList, tempVM)
+		}
+	}
+	if testOptions.windowsVMs != "" {
+		vms := strings.Split(testOptions.windowsVMs, ",")
+		for _, vm := range vms {
+			t.Logf("Get info for Windows VM: %s", vm)
+			tempVM := getWindowsVMInfo(t, data, vm)
+			vmList = append(vmList, tempVM)
+		}
+	}
+	t.Logf("TestVMAgent setup")
+	for i, vm := range vmList {
+		stopAntreaAgent(t, data, vm)
+		t.Logf("Creating ExternalNode for VM: %s", vm.nodeName)
+		en, err := createExternalNodeCRD(data, vm.nodeName, vm.ifName, vm.ip)
+		require.NoError(t, err, "Failed to create ExternalNode")
+		vmList[i].eeName, err = externalnode.GenExternalEntityName(en)
+		require.NoError(t, err, "Failed to generate ExternalEntity Name for ExternalNode %s", en.Name)
+		startAntreaAgent(t, data, vm)
+	}
+	return vmList, nil
+}
+
+// teardownVMAgentTest deletes ExternalNode, verifies ExternalEntity is deleted
+// and verifies uplink configuration is restored.
+func teardownVMAgentTest(t *testing.T, data *TestData, vmList []vmInfo) {
+	verifyUpLinkAfterCleanup := func(vm vmInfo) {
+		err := wait.PollImmediate(10*time.Second, 1*time.Minute, func() (done bool, err error) {
+			var tempVM vmInfo
+			if vm.osType == "Linux" {
+				tempVM = getVMInfo(t, data, vm.nodeName)
+			} else {
+				tempVM = getWindowsVMInfo(t, data, vm.nodeName)
+			}
+			if vm.ifName != tempVM.ifName {
+				t.Logf("Retry, unexpected uplink interface name, expected %s, got %s", vm.ifName, tempVM.ifName)
+				return false, nil
+			}
+			if vm.ip != tempVM.ip {
+				t.Logf("Retry, unexpected uplink IP, expected %s, got %s", vm.ip, tempVM.ip)
+				return false, nil
+			}
+			return true, nil
+		})
+		assert.NoError(t, err, "Failed to verify uplink configuration after cleanup")
+	}
+	t.Logf("TestVMAgent teardown")
+	for _, vm := range vmList {
+		if err := data.crdClient.CrdV1alpha1().ExternalNodes(namespace).Delete(context.TODO(), vm.nodeName, metav1.DeleteOptions{}); err != nil {
+			t.Logf("Failed to delete ExternalNode, err: %v", err)
```

Comment:
what about this? https://github.com/antrea-io/antrea/pull/3824#discussion_r941208097

Reply:
synced offline and updated the PR.